### ISSUE TITLE
Make expect in Bloc_Test a builder that returns an iterable

### ISSUE THREAD

**Issue by petercinibulk** [felangel/bloc#2010]
**Is your feature request related to a problem? Please describe.**
When using bloc_test, you cannot use global variables that are set in the setup function of a test group inside the expect field.

**Describe the solution you'd like**
Make the blocTest expect field a builder that returns an iterable.

```
blocTest(
    'emits [1] when CounterEvent.increment is added',
    build: () => CounterBloc(),
    act: (bloc) => bloc.add(CounterEvent.increment),
    expect: [1],
  );
```
into
```
blocTest(
    'emits [1] when CounterEvent.increment is added',
    build: () => CounterBloc(),
    act: (bloc) => bloc.add(CounterEvent.increment),
    expect: () => [1],
  );
```
**Describe alternatives you've considered**
Listen inside the build function and save states globally. Then you could check states in the verify builder. This would be very redundant though.

**Additional context**

**Comment by felangel** [comment]
Hi @petercinibulk 👋 
Thanks for opening an issue!

I will discuss this with the rest of the team but I'm open to making this change as part of the v8.0.0 release 👍

**Comment by SEGVeenstra** [comment]
Ran into the same issue yesterday.
Because the expect is called before build is ran, you cannot initialise properties in the build function or even in the setUp().

```dart
group('Sample', () {
  SampleItem item;

  setUp((){                                         // 2nd
    item = SampleItem("label");
  });

  blocTest(
    'Sample test',
    build: () {                                     // 3rd

      final service = MockSampleService();
      when(service.item).thenReturn(item);

      return SampleBloc(service ,SampleInitial());
    },
    act: (bloc) => bloc.add(SampleGetItemEvent()),
    expect: [SampleLoading(), SampleLoaded(item)], // 1st (item will be null)
  );
});
```
To make this workable you have to either create globals or creator functions to get same or equal instances:
```dart
group('Sample', () {
  final item = SampleItem("label");
  // OR
  SampleItem getItem() => SampleItem("label");

  ...
});
```
This is far from ideal, OP's suggestion was exactly my idea to solve this. Turn expect into a builder.

**Comment by felangel** [comment]
This is addressed by #1733  and will be available in v8.0.0-nullsafety.0 👍

**Comment by felangel** [comment]
This is currently available in v8.0.0-nullsafety.0 👍
